### ISSUE TITLE
Mask shift amounts to prevent undefined behavior

### ISSUE THREAD

**Issue by marbre** [iree-org/iree#4867]
https://github.com/google/iree/blob/dd85a5b4068a6faeb810f2e24e69e2429749da3e/iree/vm/ops.h#L175-L183

We want to mask these shift amounts to prevent undefined behavior - for example, a shift left of amount=100 will trigger undefined behavior.

from the C spec:
```
  If the value of the right operand is negative, or equal to or greater
  than the number of bits in the promoted left operand, the behavior is
  undefined.
```

Super nice examples: https://wiki.sei.cmu.edu/confluence/display/c/INT34-C.+Do+not+shift+an+expression+by+a+negative+number+of+bits+or+by+greater+than+or+equal+to+the+number+of+bits+that+exist+in+the+operand

We can use uint8_t here for amount so that we don't have to worry about negative shift amounts.
Once we have the unsigned shift amount we can mask `(amount & 63)` (and `(amount & 31)` for the 32-bit variants) and then only need to check for 0 (only perform the shift if `(amount & 63)` != 0).

Something like:
```c
  return (amount & 63) ? (int64_t)(operand << (amount & 63)) : operand;
```

I know the code didn't do this correctly before so if you don't want to fix it here we can file an issue and do it in a followup - my ubsan sense is tingling and indicating that we'll end up with fuzzing producing some errors here and we may as well make it right while we remember :)

Tests that shift by 0 and >32/>64 may be hard to author (as I think we check for these in IR and fold them away), but may be worth a shot.

_Originally posted by @benvanik in https://github.com/google/iree/pull/4863#discussion_r577724935_

**Comment by simon-camp** [comment]
@benvanik Can this be closed with the changes from [this commit](https://github.com/google/iree/commit/b8c13c77fea5cbe9473a9a9b347021e8fc66fb79)?

**Comment by benvanik** [comment]
Fixed in #5864.
